### PR TITLE
LPS-189207 (Draft)

### PR DESCRIPTION
**Pull request by kevhlee**
__Ticket:__ <https://liferay.atlassian.net/browse/LPS-189207>

Hi @achaparro, this is a draft of the partitioning approach we discussed for allowing migration of Quartz data from a virtual instance. See this comment to get an overview of how the approach needs to be implemented: <https://liferay.atlassian.net/browse/LPS-189207?focusedCommentId=2473159>.

There is one thing I have not fully addressed with this approach, which is knowing when to shut down a scheduler instance when its company is deleted. At the moment, in `QuartzSchedulerEngine`, we are able to know when a new company was created (when we see a new `companyId` we have not seen before). However, `QuartzSchedulerEngine` currently does not know when a company is deleted. I was wondering if there was an efficient way to do this. For example, is there some kind of listener or hook Liferay provides that allows us to know when a company is updated? Would having a `PortalInstanceLifecycleListener` work?

Let me know if you have any questions.

### CODE REVIEW

File: modules/apps/portal-scheduler/portal-scheduler-quartz/src/main/java/com/liferay/portal/scheduler/quartz/internal/QuartzSchedulerEngine.java [hunk 1]
```diff
@@ -796,20 +842,64 @@ private void _updateJobState(
...
+		public ConnectionProviderImpl(long companyId) {
+			_companyId = companyId;
+		}
+
+		@Override
+		public Connection getConnection() {
+			Connection connection = null;
+
+			try (SafeCloseable safeCloseable =
+					CompanyThreadLocal.setWithSafeCloseable(_companyId)) {
```
Comment (achaparro) [inline modules/apps/portal-scheduler/portal-scheduler-quartz/src/main/java/com/liferay/portal/scheduler/quartz/internal/QuartzSchedulerEngine.java:856]: @kevhlee if we use SafeClosable here we won't keep the CompanyThreadLocal set when we create the statements and set the catalog based on the CompanyThreadLocal in com.liferay.portal.db.partition.DBPartitionUtil#_getConnectionWrapper

So we should use just com.liferay.portal.kernel.security.auth.CompanyThreadLocal#setCompanyId

But we have to be sure this java thread is only use for one company scheduler, can you confirm that?

Comment (kevhlee) [reply modules/apps/portal-scheduler/portal-scheduler-quartz/src/main/java/com/liferay/portal/scheduler/quartz/internal/QuartzSchedulerEngine.java:856]: Hey @achaparro, I will try to confirm.

**Comment by achaparro** [comment]
@kevhlee regarding your question about deleting a company event, we should use something like Search uses to remove the index:
com.liferay.portal.search.internal.instance.lifecycle.SearchIndexPortalInstanceLifecycleListener

**Comment by achaparro** [comment]
@kevhlee I also think that we should only create company schedulers and custom ConnectionProvider when db partition is enabled. I don't see the benefit of using when db partition is not enabled since we only have one database schema. Let me know if there is an important benefit of using it that I am missing.

We can still have the map but with one single element when db partition is not enabled. In the getters we can just return the default names or provider when db partition is not enabled.

**Comment by kevhlee** [comment]
Hi @achaparro, thanks for taking a look! I will send a new draft, taking your comments into account.

Regarding setting `CompanyThreadLocal`, assuming only a single unique thread accesses each connection provider, would this suffice?:

```java
		@Override
		public Connection getConnection() {
			CompanyThreadLocal.setCompanyId(_companyId);

			Connection connection = null;

			try {
				DataSource dataSource = InfrastructureUtil.getDataSource();

				connection = dataSource.getConnection();
			}
			catch (Exception exception) {
				_log.error(exception);
			}

			return connection;
		}
```

File: modules/apps/portal-scheduler/portal-scheduler-quartz/src/main/java/com/liferay/portal/scheduler/quartz/internal/QuartzSchedulerEngine.java (hunk 1 above)
Comment (kevhlee) [reply modules/apps/portal-scheduler/portal-scheduler-quartz/src/main/java/com/liferay/portal/scheduler/quartz/internal/QuartzSchedulerEngine.java:856]: Hey @achaparro, regarding your question:

>But we have to be sure this java thread is only used for one company scheduler, can you confirm that?

That should not be a problem based on what I found. So, there are multiple threads for each persisted scheduler instance (e.g. one that handles misfires: `MisfireHandler`) and they will never be used for other persisted scheduler instances. So using just `CompanyThreadLocal.setCompanyId()` should be sufficient.

```log
2023-10-04 01:47:36.472 ERROR [PersistedQuartzSchedulerInstance@20001_QuartzSchedulerThread][QuartzSchedulerEngine:838] Current thread: Thread[PersistedQuartzSchedulerInstance@20001_QuartzSchedulerThread,5,QuartzScheduler:PersistedQuartzSchedulerInstance@20001]
2023-10-04 01:47:38.871 ERROR [PersistedQuartzSchedulerInstance@32675_QuartzSchedulerThread][QuartzSchedulerEngine:838] Current thread: Thread[PersistedQuartzSchedulerInstance@32675_QuartzSchedulerThread,5,QuartzScheduler:PersistedQuartzSchedulerInstance@32675]
2023-10-04 01:48:12.540 ERROR [QuartzScheduler_PersistedQuartzSchedulerInstance@20001-NON_CLUSTERED_MisfireHandler][QuartzSchedulerEngine:838] Current thread: Thread[QuartzScheduler_PersistedQuartzSchedulerInstance@20001-NON_CLUSTERED_MisfireHandler,5,main]
2023-10-04 01:48:12.579 ERROR [QuartzScheduler_PersistedQuartzSchedulerInstance@32675-NON_CLUSTERED_MisfireHandler][QuartzSchedulerEngine:838] Current thread: Thread[QuartzScheduler_PersistedQuartzSchedulerInstance@32675-NON_CLUSTERED_MisfireHandler,5,main]
```

**Comment by achaparro** [comment]
Hi @kevhlee,

Yes, that's it :-)

**Pull request closed without merging**